src/Servers/IIS/IIS/src/Core/IISHttpContext.IO.cs

```diff
@@ -186,9 +189,17 @@ private async Task WriteBody(bool flush = false)
 
         internal void AbortIO(bool clientDisconnect)
         {
-            if (Interlocked.CompareExchange(ref _requestAborted, 1, 0) != 0)
+            var shouldScheduleCancellation = false;
+
+            lock (_abortLock)
             {
-                return;
+                if (_requestAborted)
+                {
+                    return;
+                }
+
+                shouldScheduleCancellation = _abortedCts != null;
+                _requestAborted = true;
             }
 
             if (clientDisconnect)
```

Comment:
What is the clientDisconnect parameter about anyway? Should it influence whether the RequestAborted token fires?

Reply:
It represents whether the client was the one to initiate the abort or not. This would be true if we called WriteAsync and it threw a ConnectionResetException, and false if the server initiated the abort (ctx.Abort()). In both scenarios I believe we should try to cancel the abortedCts if it exists.

I think no matter what, we want RequestAborted to fire if the client or server disconnected.